File: PAP2324/src/util/Stack.java
package util;

/**
 * Stack générique ayant une taille maximum pré-définie 
 * lors de la construction
 * 
 * @param <T>
 */
public class Stack<T> {
	private final int tailleMax;
	private int sommet;
	private T[] v;

	/**
	 * Constructeur 
	 * 
	 * @param tailleMax : taille maximum du stack
	 */
	@SuppressWarnings("unchecked")
	public Stack(int tailleMax) {
		this.tailleMax = tailleMax;
		this.sommet = 0;
		this.v = (T[]) new Object[tailleMax];
	}

	/**
	 * Ajoute l'élément au stack
	 * 
	 * @param elem
	 */
	public void push(T elem) {
		assert sommet < tailleMax : "Le stack est plein";
		v[sommet] = elem;
		sommet++;
	}

	/**
	 * supprime l'élément au sommet du stack et le renvoie
	 * 
	 * @return l'élément au sommet du stack
	 */
	public T pop() {
		assert sommet > 0 : "Le stack est vide";
		sommet--;
		T tmp = v[sommet];
		v[sommet] = null;// efface l'adresse
		return tmp;
	}

	/**
	 * retourne l'élément au sommet du stack
	 * 
	 * @return l'élément au sommet du stack
	 */
	public T top() {
		assert sommet > 0 : "Le stack est vide";
		return v[sommet - 1];
	}

	/**
	 * Indique si le stack est vide
	 * 
	 * @return
	 */
	public boolean empty() {
		return sommet == 0;
	}

	@Override
	public String toString() {
		return "Stack [tailleMax=" + tailleMax + ", nbElem=" + sommet + "]";
	}

}

File: PAP2324/python/stack.py
class StackArray:
    ''' Stack implémenté via une liste Python'''
    def __init__(self):
        self.__v=[]
        
    def push(self,elem):
        self.__v.append(elem)
        
    def pop(self):
        assert len(self.__v) != 0 ; "Stack Underflow"
        elem=self.__v[-1]
        del(self.__v[-1])
        return elem
    
    def top(self):
        assert len(self.__v) != 0 ; "Stack Underflow"
        return self.__v[-1]
    
    def empty(self):
        return len(self.__v)==0
 
class StackListe:
    ''' Stack implémenté via une liste chaînée'''
    def __init__(self):
        self.__sommet = None
        
    
if __name__=="__main__":
    s1= StackArray()
    s1.push(10)
    s1.push(20)
    print(s1.empty())
    print(s1.pop())
    print(s1.top())
    print(s1.pop())
    print(s1.empty())
    s1 = None # libère le stack pour le GC

